lib/web/terminal.go

```diff
@@ -317,6 +317,63 @@ func (t *TerminalHandler) issueSessionMFACerts(tc *client.TeleportClient, ws *we
 		return trace.Wrap(err)
 	}
 
+	mfaEncode := func(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error) {
+		chalEnc, err := json.Marshal(chal)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		envelope := &Envelope{
+			Version: defaults.WebsocketVersion,
+			Type:    envelopeType,
+			Payload: string(chalEnc),
+		}
+		envelopeBytes, err := proto.Marshal(envelope)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return envelopeBytes, nil
+	}
+
+	mfaDecode := func(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error) {
+		envelope := &Envelope{}
+		if err = proto.Unmarshal(bytes, envelope); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		var mfaResponse *authproto.MFAAuthenticateResponse
+
+		// Convert from JSON to proto.
+		switch envelopeType {
+		case defaults.WebsocketWebauthnChallenge:
+			var webauthnResponse wanlib.CredentialAssertionResponse
+			if err := json.Unmarshal([]byte(envelope.Payload), &webauthnResponse); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			mfaResponse = &authproto.MFAAuthenticateResponse{
+				Response: &authproto.MFAAuthenticateResponse_Webauthn{
+					Webauthn: wanlib.CredentialAssertionResponseToProto(&webauthnResponse),
+				},
+			}
+
+		default:
+			var u2fResponse u2f.AuthenticateChallengeResponse
+			if err := json.Unmarshal([]byte(envelope.Payload), &u2fResponse); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			mfaResponse = &authproto.MFAAuthenticateResponse{
+				Response: &authproto.MFAAuthenticateResponse_U2F{
+					U2F: &authproto.U2FResponse{
+						KeyHandle:  u2fResponse.KeyHandle,
+						ClientData: u2fResponse.ClientData,
+						Signature:  u2fResponse.SignatureData,
+					},
+				},
+			}
+		}
+
+		return mfaResponse, nil
+	}
```

Comment:
nit:
What would you say about defining ecplicty MFAChalangeEncoder and MFAChallangeDecore interfaces instaed of using inline function and having the  `MFAChalangeEncDec` and MFAChallnageRDSEncDec types:

```go
   type MFAChalangeEncDec interface {
      Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error)
      Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)
   }
```
and typen just defining the Proto RDS MFA types like:
```go
type MFAChallnageProtoEncDec struct {
}

func (m *MFAChallnageProtoDecoder) Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error) {
  ...
}

func (m *MFAChallnageProtoDecoder)  Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)([]byte, error) {
  ...
}
```
```go
type MFAChallnageRDSEncDec struct {
}

func (m *MFAChallnageRDSEncDec) Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error) {
  ...
}

func (m *MFAChallnageRDSEncDec)  Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)([]byte, error) {
  ...
}
```
and during the `promptMFAChallenge` call we can just pass a proper type:
`promptMFAChallenge(ws, t.wsLock, &MFAChallnageRDSEncDec{})`

Reply:
That's exactly what I was thinking. I'm going to put this change together now and update the PR.

Reply:
Good idea.